pricegraph/wasm/src/lib.rs

```diff
@@ -0,0 +1,37 @@
+//! This crate provides a thin WASM-compatible wrapper around the `pricegraph`
+//! crate and can be used for estimating prices for a given orderbook.
+
+use pricegraph::{Orderbook, TokenId, TokenPair};
+use wasm_bindgen::prelude::*;
+
+/// A graph representation of a complete orderbook.
+#[wasm_bindgen]
+pub struct PriceEstimator {
+    orderbook: Orderbook,
+}
+
+#[wasm_bindgen]
+impl PriceEstimator {
+    /// Creates a `PriceEstimator` instance by reading an orderbook from encoded
+    /// bytes. Returns an error if the encoded orders are invalid.
+    #[wasm_bindgen(constructor)]
+    pub fn new(bytes: &[u8]) -> Result<PriceEstimator, JsValue> {
+        console_error_panic_hook::set_once();
+
+        let mut orderbook = Orderbook::read(bytes).map_err(|err| err.to_string())?;
+        orderbook.reduce_overlapping_orders();
+
+        Ok(PriceEstimator { orderbook })
+    }
+
+    /// Estimates price for the specified trade. Returns `undefined` if the
+    /// volume cannot be fully filled.
+    #[wasm_bindgen(js_name = "estimatePrice")]
+    pub fn estimate_price(&self, buy: TokenId, sell: TokenId, volume: f64) -> Option<f64> {
```

Comment:
This is too advanced for my rust knowledge, but even risking on making a silly question, why do you return an `Option` here?

Maybe is because it needs to warm up the estimation, but if that's the case, and given that you use `unwrap` in line `28` and that you said you cannot handle panics and they are just console.logged. 

Shouldn't we handle the None case so it doesn't panics, and we return that wrapped result you mentioned in the comment?

Reply:
>  but even risking on making a silly question

No such thing!

> This is too advanced for my rust knowledge, but even risking on making a silly question, why do you return an Option here?

`Option` in rust gets converted to a nullable type in TS, so its like it returns `number?`. This is needed because estimates aren't always possible since there isn't always a "path" from the buy token to the sell token that allows for trading. The most trivial example of this is if you are trying to estimate a price for buying A and selling B, but there are no open orders that are buying B - so it doesn't matter how much you offer, your order won't get matched.

> Shouldn't we handle the None case so it doesn't panics, and we return that wrapped result you mentioned in the comment?

`unwrap` is bad because panics can't be caught in WASM since there is no stack unwinding (which is how they are implemented in Rust) and instead they function more like a C `abort` which can't be caught. That being said, Rust allows for "panic hooks" in WASM that run arbitrary code right before a panic happens and the code aborts. The `unwrap` you saw is in the integration tests module, where the test runner takes care to add a panic hook that signals the test has failed in case it panics which is what we want in that case.

Reply:
Ahhhh! okay, it's from the test module. My concern was just that I thought it was unwrapping in the module, so if there's price it returns the price, and if it doesn't have a price, it was panicking (so console.error)

Thanks for explaining!